The aim of this project is calculate and compare contemporary (nucleotide diversity or pi) and historical (temporal change in allele frequencies) effective population size in Anopheles coluzzii and Anopheles gambiae using autosomes and sex chromosomes. To estimate the population size of the Anopheles mosquito using both methods. I have used genetic data from the MalariaGEN database (The Anopheles gambiae 1000 Genomes Consortium, 2021, https://www.malariagen.net/data/ag1000g-phase3-snp). However, it is not possible to estimate directly the population size using the enormous genetic information so several filters must be applied to select only the loci that are of interest for our study. For this purpose, some codes have been developed in Phyton and R.
First, from the database of MalariaGEN, some filters must be applied to the genomic data to obtain databases with useful information. This part has been done by Hui, Tin-Yu and its script in Python using Jupyter Notebook can be found here: https://github.com/tinyuhui/AG3. From this database, you can download a hdf5 file which contains four types of information: the genotype of each base pair or locus, the position of each base pair or locus, the sample id of each sampled mosquito and the metadata. In addition, the genomic data has been filtered. The loci included in the TCAF method are SNPs (without fixed sites) that have the following filters: biallelic, intergenic, without missing sites, MAF > 0.05 (minor allele frequency must be higher than 5 %). Using the nucleotide diversity method you it is also included the fixed sites. For this reason, you obtain files with different information depending on the method. Finally, you can select the chromosome which you want to study (X, 3R, 3L, 2R, 2L). However, in the script we will only use chr X as an example.

