The aim of this project is calculate and compare contemporary (nucleotide diversity or pi) and historical (temporal change in allele frequencies) effective population size in Anopheles coluzzii and Anopheles gambiae using autosomes and sex chromosomes. Therefore, the population size of the Anopheles mosquito is estimated using both methods. I have used genetic data from the MalariaGEN database (The Anopheles gambiae 1000 Genomes Consortium, 2021, https://www.malariagen.net/data/ag1000g-phase3-snp). However, it is not possible to estimate directly the population size using the enormous genetic information so several filters must be applied to select only the loci that are of interest for our study. For this purpose, some codes have been developed in Phyton and R.
First, from the database of MalariaGEN, some filters must be applied to the genomic data to obtain databases with useful information. This part has been done by Brenas, Jon and Hui, Tin-Yu and its script in Python using Jupyter Notebook can be found here: https://github.com/tinyuhui/AG3. From this database, you can download a hdf5 file which contains four types of information: the genotype of each base pair or locus, the position of each base pair or locus, the sample id of each sampled mosquito and the metadata. In addition, the genomic data has been filtered. The loci included in the TCAF (temporary change in allele frequencies) method are SNPs (without fixed sites) that have the following filters: biallelic, intergenic, without missing sites. Using the nucleotide diversity method you it is also included the fixed sites. For this reason, you obtain files with different information depending on the method. Finally, you can select the chromosome which you want to study (X, 3R, 3L, 2R, 2L). However, in the script we will only use chr X as an example.
When the hdf5 file and the metadata file are downloaded, the first one has to be recoded. This means that genotypes are coded like this: 00, 01 and 11, but we need to convert them to 0, 1 and 2 in order to calculate properly the allele frequencies. To do this we used the recode_genotypes_X.R and Rside.c files, which can be found in this repository. The first one loads the hdf5 file and export two files: genomic_X_TCAF.RData and genomic_X_pi.RData, which can be also found in this repository. The second one is zipped in the file genomic_X_pi.zip due to its large size. Finally, recode_genotypes_X.R and Rside.c have been coded by Hui, Tin-Yu and the second file has to be compiled in order to obtain the Rside.dll file, which is the one that we need in recode_genotypes_X.R file.
When we obtain the genomic_X_TCAF.R and genomic_X_pi.RD files, we need to apply more filters but in this case will be done in R and all the files from now are located in this repository except of the estimation of population size using TCAF method. First we will explain the first file. With genomic_X_TCAF.R, we load two other files: genomic_X_TCAF.RData and metadata.csv. The first one contains genotype, position and sample_ID and the second one information about the sample ID. First, we filter the sample_id obtaining only the samples coluzzii 2012, 2014 and gambiae 2012, 2014 using the metadata.csv file. With these sample_ids we can filter the genotypes and obtaining four genotypes: genotype_col_2012, genotype_col_2014, genotype_gamb_2012 and genotype_gamb_2014. Then we estimate the allele frequencies using the genotypes and their sample sizes (number of sampled mosquitoes). 
, MAF > 0.05 (minor allele frequency must be higher than 5 %)

